eth_account/_utils/structured_data/hashing.py

```diff
@@ -0,0 +1,279 @@
+            hashed_value = keccak(primitive=value)
+            yield "bytes32", hashed_value
+        elif field["type"] in types:
+            # This means that this type is a user defined type
+            hashed_value = keccak(primitive=encode_data(field["type"], types, value))
+            yield "bytes32", hashed_value
+        elif is_array_type(field["type"]):
+            # Get the dimensions from the value
+            array_dimensions = get_array_dimensions(value)
+            # Get the dimensions from what was declared in the schema
+            parsed_type = parse(field["type"])
+            for i in range(len(array_dimensions)):
+                if len(parsed_type.arrlist[i]) == 0:
+                    # Skip empty or dynamically declared dimensions
+                    continue
+                if array_dimensions[i] != parsed_type.arrlist[i][0]:
+                    # Dimensions should match with declared schema
+                    raise TypeError(
+                        "Array data `{0}` has dimensions `{1}` whereas the "
+                        "schema has dimensions `{2}`".format(
+                            value,
+                            array_dimensions,
+                            tuple(map(lambda x: x[0], parsed_type.arrlist)),
+                        )
+                    )
+
+            array_items = flatten_multidimensional_array(value)
+            array_items_encoding = [
+                encode_data(parsed_type.base, types, array_item)
+                for array_item in array_items
+            ]
+            concatenated_array_encodings = ''.join(array_items_encoding)
+            hashed_value = keccak(concatenated_array_encodings)
+            yield "bytes32", hashed_value
+        else:
+            # First checking to see if type is valid as per abi
+            if not is_valid_abi_type(field["type"]):
+                raise TypeError(
+                    "Received Invalid type `{0}` in the struct `{1}`".format(
+                        field["type"],
+                        primary_type,
+                    )
+                )
+
+            # Next see if the data fits the specified encoding type
+            if is_encodable(field["type"], value):
+                # field["type"] is a valid type and this value corresponds to that type.
+                yield field["type"], value
+            else:
+                raise TypeError(
+                    "Value of `{0}` ({2}) in the struct `{1}` is of the type `{3}`, but expected "
+                    "{4} value".format(
+                        field["name"],
+                        primary_type,
+                        value,
+                        type(value),
+                        field["type"],
+                    )
+                )
+
+
+def encode_data(primaryType, types, data):
+    data_types_and_hashes = _encode_data(primaryType, types, data)
+    data_types, data_hashes = zip(*data_types_and_hashes)
```

Comment:
This will fail if `data_types_and_hashes` is empty.  Something to be aware of with this pattern that typically has to be special cased.

Reply:
@pipermerriam such a case won't occur. This is because `_encode_data` function returns atleast one `2 length tuple`, which would be `("bytes32", hash_struct_type(primary_type, types))`